### PR TITLE
[SPARK-18619][ML] Make QuantileDiscretizer/Bucketizer/StringIndexer/RFormula inherit from HasHandleInvalid

### PR DESCRIPTION
**Pull request by zhengruifeng**
## What changes were proposed in this pull request?
1, HasHandleInvaild support override
2, Make QuantileDiscretizer/Bucketizer/StringIndexer/RFormula inherit from HasHandleInvalid

## How was this patch tested?
existing tests

[JIRA](https://issues.apache.org/jira/browse/SPARK-18619)

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala [hunk 1]
```diff
@@ -36,7 +36,8 @@ import org.apache.spark.sql.types.{DoubleType, StructField, StructType}
  */
 @Since("1.4.0")
 final class Bucketizer @Since("1.4.0") (@Since("1.4.0") override val uid: String)
-  extends Model[Bucketizer] with HasInputCol with HasOutputCol with DefaultParamsWritable {
+  extends Model[Bucketizer] with HasHandleInvalid with HasInputCol with HasOutputCol
+    with DefaultParamsWritable {
```
Comment (facaiy) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala]: How about aligning `with` with `extends`?

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala [hunk 2]
```diff
@@ -460,16 +460,16 @@ object LinearRegression extends DefaultParamsReadable[LinearRegression] {
   val MAX_FEATURES_FOR_NORMAL_SOLVER: Int = WeightedLeastSquares.MAX_NUM_FEATURES
 
   /** String name for "auto". */
-  private[regression] val AUTO = "auto"
+  private[regression] val Auto = "auto"
```
Comment (facaiy) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: Is `solver` related with `handleInvalid`? It seems a little confused.

File: python/pyspark/ml/feature.py [hunk 3]
```diff
@@ -3058,26 +3035,37 @@ class RFormula(JavaEstimator, HasFeaturesCol, HasLabelCol, JavaMLReadable, JavaM
                                    "RFormula drops the same category as R when encoding strings.",
                                    typeConverter=TypeConverters.toString)
 
+    handleInvalid = Param(Params._dummy(), "handleInvalid", "how to handle invalid entries. " +
```
Comment (facaiy) [inline python/pyspark/ml/feature.py]: If I understand correctly, `handleInvalid` has been declared in `HasHandleInvalid` interface, right?

File: mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala [hunk 4]
```diff
@@ -36,7 +36,8 @@ import org.apache.spark.util.collection.OpenHashMap
 /**
  * Base trait for [[StringIndexer]] and [[StringIndexerModel]].
  */
-private[feature] trait StringIndexerBase extends Params with HasInputCol with HasOutputCol {
+private[feature] trait StringIndexerBase extends Params with HasHandleInvalid with HasInputCol
```
Comment (facaiy) [inline mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala]: Perhaps it is better to break the line at `extends`.

File: mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala [hunk 5]
```diff
@@ -74,16 +74,12 @@ private[feature] trait QuantileDiscretizerBase extends Params
    */
   // TODO: SPARK-18619 Make QuantileDiscretizer inherit from HasHandleInvalid.
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala]: Remove this line.

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala (hunk 2 above)
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: @facaiy It's not related to this PR, just addressed other small issues by the way.

File: python/pyspark/ml/feature.py [hunk 6]
```diff
@@ -3058,26 +3035,37 @@ class RFormula(JavaEstimator, HasFeaturesCol, HasLabelCol, JavaMLReadable, JavaM
...
+                          "Options are 'skip' (filter out rows with invalid values), " +
+                          "'error' (throw an error), or 'keep' (put invalid data in a special " +
+                          "additional bucket, at index numLabels).",
+                          typeConverter=TypeConverters.toString)
+
     @keyword_only
     def __init__(self, formula=None, featuresCol="features", labelCol="label",
-                 forceIndexLabel=False, stringIndexerOrderType="frequencyDesc"):
+                 forceIndexLabel=False, handleInvalid="error",
+                 stringIndexerOrderType="frequencyDesc"):
```
Comment (yanboliang) [inline python/pyspark/ml/feature.py]: We should append new added argument in case to break users' code.

File: python/pyspark/ml/feature.py (hunk 3 above)
Comment (zhengruifeng) [inline python/pyspark/ml/feature.py]: @facaiy We have to override it to keep the pydoc in line with scala side.

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala (hunk 2 above)
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: @facaiy We had some offline discussion, and decided to fix some small issues in this PR.

**Comment by yanboliang** [comment]
LGTM, merged into master. Thanks.

**Pull request closed without merging**